src/net/server/mod.rs

```diff
@@ -210,6 +242,267 @@ where
+        .filter_map(|peer| peer.certificate.as_ref())
+    {
+        // Note that this uses `webpki::TrustAnchor::try_from_cert_der`, which *does not* validate
+        // the certificate. That is not required for security, but might be desirable to flag
+        // configuration errors.
+        trusted_certs.add(cert)?;
+    }
+    let verifier = AllowAnyAnonymousOrAuthenticatedClient::new(trusted_certs);
+
+    let mut config = RustlsServerConfig::builder()
+        .with_safe_defaults()
+        .with_client_cert_verifier(verifier)
+        .with_single_cert(cert, key)?;
+
+    config.alpn_protocols = vec![b"h2".to_vec(), b"http/1.1".to_vec()];
+
+    Ok(RustlsConfig::from_config(Arc::new(config)))
+}
+
+/// Axum `Extension` indicating the authenticated remote helper identity, if any.
+//
+// Presence or absence of authentication is indicated by presence or absence of the extension. Even
+// at some inconvenience (e.g. `MaybeExtensionExt`), we avoid using `Option` within the extension,
+// to avoid possible confusion about how many times the return from `req.extensions().get()` must be
+// unwrapped to ensure valid authentication.
+#[derive(Clone, Copy, Debug)]
+struct ClientIdentity(pub HelperIdentity);
+
+impl Deref for ClientIdentity {
+    type Target = HelperIdentity;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+/// `Accept`or that sets an axum `Extension` indiciating the authenticated remote helper identity.
+#[derive(Clone)]
+struct ClientCertRecognizingAcceptor {
+    inner: RustlsAcceptor,
+    network_config: Arc<NetworkConfig>,
+}
+
+impl ClientCertRecognizingAcceptor {
+    fn new(inner: RustlsAcceptor, network_config: NetworkConfig) -> Self {
+        Self {
+            inner,
+            network_config: Arc::new(network_config),
+        }
+    }
+
+    // This can't be a method (at least not that takes `&self`) because it needs to go in a 'static future.
+    fn identify_client(
+        network_config: &NetworkConfig,
+        cert_option: Option<&Certificate>,
+    ) -> Option<ClientIdentity> {
+        let Some(cert) = cert_option else {
+            return None;
+        };
+        // Fast and easy -- look for an exact match of certificate
+        for (id, peer) in network_config.enumerate_peers() {
+            if peer.certificate.as_ref() == Some(cert) {
+                return Some(ClientIdentity(id));
+            }
+        }
+        // No exact match, look for a trust relationship
+        let end_entity_cert = match EndEntityCert::try_from(cert.as_ref()) {
+            Ok(cert) => cert,
+            Err(err) => {
+                error!("unable to parse client certificate: {err}");
+                return None;
+            }
+        };
+        for (id, peer) in network_config.enumerate_peers() {
+            let Ok(name) = DnsNameRef::try_from_ascii_str(peer.url.host().unwrap()) else {
+                error!("peer {} has an invalid hostname", peer.url.host().unwrap());
```

Comment:
This is a case where configuration is bad, so it would be OK to `panic!()`.  The only reason I might prefer this is that the error tends to appear pretty late in the whole process (after the network is setup and a query comes in).